bench/bench.ml

```diff
@@ -0,0 +1,487 @@
+     Number of bindings: %d@\n\
+     Log size: %d@\n\
+     Seed: %d@\n\
+     Metrics: %b" config.key_size config.value_size config.nb_entries
+    config.log_size config.seed config.with_metrics
+
+let cleanup root =
+  let files = [ "data"; "log"; "lock"; "log_async"; "merge" ] in
+  List.iter
+    (fun (b : Index.suite_elt) ->
+      let dir = root // b.name // "index" in
+      List.iter
+        (fun file ->
+          let file = dir // file in
+          if Sys.file_exists file then Unix.unlink file)
+        files)
+    Index.suite
+
+let init config =
+  Printexc.record_backtrace true;
+  Random.init config.seed;
+  if config.with_metrics then (
+    Metrics.enable_all ();
+    Metrics_gnuplot.set_reporter ();
+    Metrics_unix.monitor_gc 0.1 );
+  bindings_pool := make_bindings_pool config.nb_entries;
+  absent_bindings := make_bindings_pool config.nb_entries
+
+let print fmt (config, results) =
+  let pp_bench fmt (b, result) =
+    Format.fprintf fmt "%s@\n    @[%a@]" b.Index.synopsis Benchmark.pp_result
+      result
+  in
+  Format.fprintf fmt
+    "Configuration:@\n    @[%a@]@\n@\nResults:@\n    @[%a@]@\n" pp_config config
+    Fmt.(list ~sep:Fmt.(any "@\n@\n") pp_bench)
+    results
+
+let print_json fmt (config, results) =
+  let open Yojson.Safe in
+  let obj =
+    `Assoc
+      [
+        ("config", config_to_yojson config);
+        ( "results",
+          `Assoc
+            (List.map
+               (fun (b, result) ->
+                 (b.Index.name, Benchmark.result_to_yojson result))
+               results) );
+      ]
+  in
+  pretty_print fmt obj
+
+let run filter root seed with_metrics log_size nb_entries json =
+  let config =
+    { key_size; value_size; nb_entries; log_size; seed; with_metrics }
+  in
+  cleanup root;
+  init config;
+  let name_filter =
+    match filter with None -> fun _ -> true | Some re -> Re.execp re
+  in
+  Index.suite
+  |> schedule name_filter
+  |> List.map (fun (b : Index.suite_elt) ->
+         let name =
+           match b.dependency with None -> b.name | Some name -> name
+         in
+         ( b,
+           Index.run ~with_metrics ~nb_entries ~log_size ~root ~name b.benchmark
+         ))
+  |> fun results ->
+  Fmt.pr "%a" (if json then print_json else print) (config, results)
+
+open Cmdliner
+
+let env_var s = Arg.env_var ("INDEX_BENCH_" ^ s)
+
+let regex =
+  let parse s =
+    try Ok Re.(compile @@ Pcre.re s) with
+    | Re.Perl.Parse_error -> Error (`Msg "Perl-compatible regexp parse error")
+    | Re.Perl.Not_supported -> Error (`Msg "unsupported regexp feature")
+  in
+  let print = Re.pp_re in
+  Arg.conv (parse, print)
+
+let name_filter =
+  let doc = "A regular expression matching the names of benchmarks to run" in
+  let env = env_var "NAME_FILTER" in
+  Arg.(
+    value
+    & opt (some regex) None
+    & info [ "f"; "filter" ] ~env ~doc ~docv:"NAME_REGEX")
+
+let data_dir =
+  let doc = "Set directory for the data files" in
+  let env = env_var "DATA_DIR" in
```

Comment:
```suggestion
  let env = env_var "INDEX_DATA_DIR" in
```
I prefer having some sort of prefix here, just in case these interface with e.g. Dune invocations.

Reply:
This is already prefixed with `INDEX_BENCH_` in the definition of `env_var`.

Reply:
I see, OK.